### PR TITLE
feat(roles): add list support for roles

### LINKED ISSUE #5753: feat(roles): add support for roles list
- Add support to list all roles for a user. Roles which are equal and below given entity level should be listed
- Add support to list roles at a given entity level for a user.

### PR DESCRIPTION
**Pull request by apoorvdixit88**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Add support to 
- List available roles with info for the user
- List roles at entity level for the user

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Closes #5753 

## How did you test it?

To List roles available roles in hierarchy with info:
```
curl --location 'http://localhost:8080/user/role/v2/list' \
--header 'Authorization: Bearer JWT'
```
Response
```
[
    {
        "role_id": "merchant_operator",
        "role_name": "operator",
        "entity_type": "merchant",
        "groups": [
            "operations_view",
            "operations_manage",
            "connectors_view",
            "workflows_view",
            "analytics_view",
            "users_view",
            "merchant_details_view"
        ],
        "scope": "organization"
    },
    {
        "role_id": "merchant_developer",
        "role_name": "developer",
        "entity_type": "merchant",
        "groups": [
            "operations_view",
            "connectors_view",
            "analytics_view",
            "users_view",
            "merchant_details_view",
            "merchant_details_manage"
        ],
        "scope": "organization"
    },
    {
        "role_id": "merchant_iam_admin",
        "role_name": "iam",
        "entity_type": "merchant",
        "groups": [
            "operations_view",
            "analytics_view",
            "users_view",
            "users_manage",
            "merchant_details_view"
        ],
        "scope": "organization"
    },
    {
        "role_id": "merchant_view_only",
        "role_name": "view_only",
        "entity_type": "merchant",
        "groups": [
            "operations_view",
            "connectors_view",
            "workflows_view",
            "analytics_view",
            "users_view",
            "merchant_details_view"
        ],
        "scope": "organization"
    },
    {
        "role_id": "merchant_customer_support",
        "role_name": "customer_support",
        "entity_type": "merchant",
        "groups": [
            "operations_view",
            "analytics_view",
            "users_view",
            "merchant_details_view"
        ],
        "scope": "organization"
    },
    {
        "role_id": "merchant_admin",
        "role_name": "admin",
        "entity_type": "merchant",
        "groups": [
            "operations_view",
            "operations_manage",
            "connectors_view",
            "connectors_manage",
            "workflows_view",
            "workflows_manage",
            "analytics_view",
            "users_view",
            "users_manage",
            "merchant_details_view",
            "merchant_details_manage"
        ],
        "scope": "organization"
    },
    {
        "role_id": "org_admin",
        "role_name": "organization_admin",
        "entity_type": "organization",
        "groups": [
            "operations_view",
            "operations_manage",
            "connectors_view",
            "connectors_manage",
            "workflows_view",
            "workflows_manage",
            "analytics_view",
            "users_view",
            "users_manage",
            "merchant_details_view",
            "merchant_details_manage",
            "organization_manage"
        ],
        "scope": "organization"
    }
]
```
To list roles at entity level for invite:
```
curl --location 'http://localhost:8080/user/role/list/invite?entity_type=merchant' \
--header 'Authorization: Bearer JWT'
```
Entity type can be `organization`, `merchant` or `profile`
Response:
```
[
    {
        "role_id": "merchant_customer_support",
        "role_name": "customer_support"
    },
    {
        "role_id": "merchant_admin",
        "role_name": "admin"
    },
    {
        "role_id": "merchant_view_only",
        "role_name": "view_only"
    },
    {
        "role_id": "merchant_developer",
        "role_name": "developer"
    },
    {
        "role_id": "merchant_iam_admin",
        "role_name": "iam"
    },
    {
        "role_id": "merchant_operator",
        "role_name": "operator"
    }
]
```
To list all roles at entity level for update
```
curl --location 'http://localhost:8080/user/role/list/update?entity_type=merchant' \
--header 'Authorization: Bearer JWT'
```
Response:
```
[
    {
        "role_id": "merchant_customer_support",
        "role_name": "customer_support"
    },
    {
        "role_id": "merchant_admin",
        "role_name": "admin"
    },
    {
        "role_id": "merchant_view_only",
        "role_name": "view_only"
    },
    {
        "role_id": "merchant_developer",
        "role_name": "developer"
    },
    {
        "role_id": "merchant_iam_admin",
        "role_name": "iam"
    },
    {
        "role_id": "merchant_operator",
        "role_name": "operator"
    }
]
```
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/user_role/role.rs [hunk 1]
```diff
@@ -322,3 +322,147 @@ pub async fn update_role(
...
+            .into());
+        }
+    };
+
+    let custom_roles_map = custom_roles.into_iter().map(|role| {
+        let role_info = roles::RoleInfo::from(role);
+        role_api::MinimalRoleInfo {
+            role_id: role_info.get_role_id().to_string(),
+            role_name: role_info.get_role_name().to_string(),
+        }
```
Comment (ThisIsMani) [inline crates/router/src/core/user_role/role.rs:462]: Add entity type check and is_updatable / is_invitable.

File: crates/router/src/routes/app.rs [hunk 2]
```diff
@@ -1840,7 +1840,19 @@ impl User {
                         .route(web::get().to(get_role_from_token))
                         .route(web::post().to(create_role)),
                 )
-                .service(web::resource("/list").route(web::get().to(list_all_roles)))
+                .service(
+                    web::scope("/list")
+                        .service(web::resource("").route(web::get().to(list_all_roles)))
+                        .service(web::resource("/v2").route(web::get().to(list_roles_with_info)))
```
Comment (ThisIsMani) [inline crates/router/src/routes/app.rs:1846]: Change this to `/v2/list` if possible.

File: crates/router/src/core/user_role/role.rs (hunk 1 above)
Comment (apoorvdixit88) [reply crates/router/src/core/user_role/role.rs]: entity type is already handled in db query, added check for invitable/updatable

File: crates/router/src/core/user_role/role.rs [hunk 3]
```diff
@@ -322,3 +322,155 @@ pub async fn update_role(
...
+            role_api::RoleCheckType::Update => role_info.is_updatable(),
+        } {
+            Some(role_api::MinimalRoleInfo {
+                role_id: role_info.get_role_id().to_string(),
+                role_name: role_info.get_role_name().to_string(),
+            })
+        } else {
+            None
+        }
+    });
```
Comment (ThisIsMani) [inline crates/router/src/core/user_role/role.rs:471]: NIT: Push every thing to a vector once and then do these checks.

**Review by ThisIsMani — DISMISSED** [review]

**Review by lsampras — DISMISSED** [review]

File: crates/router/src/routes/user_role.rs [hunk 4]
```diff
@@ -286,3 +286,70 @@ pub async fn list_users_in_lineage(state: web::Data<AppState>, req: HttpRequest)
...
+pub async fn list_roles_with_info(state: web::Data<AppState>, req: HttpRequest) -> HttpResponse {
+    let flow = Flow::ListRolesV2;
+
+    Box::pin(api::server_wrap(
+        flow,
+        state.clone(),
+        &req,
+        (),
+        |state, user_from_token, _, _| role_core::list_roles_with_info(state, user_from_token),
+        &auth::DashboardNoPermissionAuth,
```
Comment (racnan) [inline crates/router/src/routes/user_role.rs:299]: `UserRead` Permission required here.

**Review by racnan — DISMISSED** [review]

File: crates/router/src/routes/user_role.rs (hunk 4 above)
Comment (apoorvdixit88) [reply crates/router/src/routes/user_role.rs:299]: Yes that would be more precise, for now let's keep this auth only.

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/routes/user_role.rs [hunk 5]
```diff
@@ -305,7 +305,7 @@ pub async fn list_updatable_roles_at_entity_level(
                 role_api::RoleCheckType::Update,
             )
         },
-        &auth::DashboardNoPermissionAuth,
+        &auth::JWTAuth(Permission::UsersRead),
```
Comment (ThisIsMani) [inline crates/router/src/routes/user_role.rs:308]: Should it be UsersWrite?

File: crates/router/src/routes/user_role.rs [hunk 6]
```diff
@@ -279,7 +279,7 @@ pub async fn list_invitable_roles_at_entity_level(
                 role_api::RoleCheckType::Invite,
             )
         },
-        &auth::DashboardNoPermissionAuth,
+        &auth::JWTAuth(Permission::UsersRead),
```
Comment (ThisIsMani) [inline crates/router/src/routes/user_role.rs:282]: Should it be UsersWrite?

**Review by ThisIsMani — APPROVED** [review]
Minor concerns.

File: crates/router/src/routes/user_role.rs (hunk 5 above)
Comment (apoorvdixit88) [reply crates/router/src/routes/user_role.rs:308]: No, it should be read only, invite function will check for write, list should be read only.

**Review by jarnura — APPROVED** [review]

File: crates/router/src/core/user_role/role.rs [hunk 7]
```diff
@@ -213,3 +213,148 @@ pub async fn update_role(
...
+            )
+            .into());
+        }
+    };
+
+    role_info_vec.extend(custom_roles.into_iter().map(roles::RoleInfo::from));
+    let list_role_info_response = role_info_vec
+        .into_iter()
+        .filter_map(|role_info| {
+            if user_role_entity >= role_info.get_entity_type() {
```
Comment (racnan) [inline crates/router/src/core/user_role/role.rs:264]: you can use [then on bool](https://doc.rust-lang.org/std/primitive.bool.html#method.then).

**Review by racnan — APPROVED** [review]

**Pull request merged**